components/ws-daemon/pkg/netlimit/netlimit.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package netlimit
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+
+	"runtime"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/kubernetes"
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/dispatch"
+	"github.com/gitpod-io/gitpod/ws-daemon/pkg/nsinsider"
+	"github.com/google/nftables"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/vishvananda/netns"
+)
+
+type ConnLimiter struct {
+	mu             sync.RWMutex
+	limited        map[string]bool
```

Comment:
How about using `sync.Map` instead of `mu`? And we can use `struct{}{}` as mark instead of `bool`
```suggestion
	limited        sync.Map
```

Reply:
Using a sync.Map would introduce a race condition because we could get a WorkspaceAdded event immediately followed by a WorkspaceUpdated event. If WorkspaceAdded is still processed then the sync.Map would not contain an entry yet and we would try to limit twice. Using a mutex serializes the updates to the limiting rules. I have updated the map to use an empty struct.

Reply:
AFAI, in that case, you said either event should be executed. What about doing a LoadOrStore first and Delete if it fails?　　e.g. https://github.com/gitpod-io/gitpod/pull/12215#discussion_r950433725
This is more efficient because the lock is not applied to the entire map.
I learned `sync.Map` from this article.
https://sreramk.medium.com/go-inside-sync-map-how-does-sync-map-work-internally-97e87b8e6bf

Reply:
The way I see it, this is not an operation that is called many thousands of times per second, but only rarely (at most 3 times per second in my experience, usually less). Any contention on the lock is likely to be short lived. I consider simplicity therefore more important than efficiency in this case.

Reply:
Make sense🙆